Added changing the casus code from one tank to another

JorgeVidal29 opened the pull request:
**Description**
Players can now copy the casus code from one tank to another

**Resolves These Issues**
Resolves #494 

**Type of change**
Check options that are relevant.

- [ ] Bug fix (fixes a bug issue)
- [x] New feature (adds functionality)
- [ ] This fix or feature will cause existing functionality to not work as expected. (Please elaborate in Description.)

**Steps to Verify Functionality**
Went to Armory
Clicked on change casus code
Selected a different tank
Clicked on copy
Went to the casus code to see the difference

**Final Checklist:**
Go down the list and check them off.

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have performed a self-review of my own code
- [x] My changes pass Flow, build without errors and (if applicable) pass Jest tests.
- [ ] This change requires a update in the design document (if applicable)

Zexdur committed:
Added changing the casus code from one tank to another

Zexdur committed:
Update Tank.js

SecondThread commented on src/tanks/Tank.js:
```diff
@@ -54,6 +53,7 @@ import {
 } from '../casus/userInteraction/CasusSpecialVariables.js';
 
 import type Battleground from '../battleground/Battleground.js';
+import ContainerBlock from '../casus/blocks/ContainerBlock.js';
```
This should be in the real imports section, not in the type imports section. (type imports get stripped out by the code we send to clients, so they don't actually cause JS cycles, that is why they are in a different section)

SecondThread commented on src/globalComponents/apiCalls/tankAPIIntegration.js:
```diff
@@ -337,6 +339,41 @@ function getTanksById(tankIds: Array<string>, onLoad:(tanks: Array<Tank>) => voi
 	)
 }
 
+function updateCasusCode(tankId: string, casusCode: ContainerBlock): void {
```
This is duplicate code from `/casus/saveCasus.js`. You should probably make saveCasus pass in an optional tankId and make it only load from a cookie if the passed in id is null. Then we only need to add ~2 lines instead of ~50

SecondThread commented on src/armory/CopyCasusCodePopup.js:
```diff
@@ -0,0 +1,95 @@
+		for(let i = 0; i < this.props.usersTanks.length; i++) {
+			if(this.props.usersTanks[i]._id === tankBeingUsedId) {
+				indexOfTankBeingUsed = i;
+				i = this.props.usersTanks.length;
```
umm, what? Is this just a sketchy `break;`?

SecondThread commented on src/armory/CopyCasusCodePopup.js:
```diff
@@ -0,0 +1,95 @@
+		}
+	}
+
+	setTankBeingCopiedFrom(tankBeingUsedId: string): void {
```
I would double check to make sure you can't just make the value of these option tags just a tank object. I was under the impression that you could. That would make pretty much this entire method one line which would be very helpful.

JorgeVidal29 replied:
If you do the value becomes [Object object] and not the actual object

SecondThread commented on src/armory/CopyCasusCodePopup.js:
```diff
@@ -0,0 +1,95 @@
+			this.setState({tankBeingCopiedFrom: null});
+			return;
+		}
+		this.setState({tankBeingCopiedFrom: this.props.usersTanks[indexOfTankBeingUsed]});
```
If, for whatever reason, you can't just pass the tank object into this method, you can still clean it up quite a bit to something like:
`this.setState({tankBeingCopiedFrom: this.props.userTanks.find(tank => tank._id == tankBeingUsedId});`

SecondThread replied:
Obviously if passing in the tanks works, you can do just:
`this.setState({tankBeingCopiedFrom: tankBeingUsed});`

SecondThread commented on src/armory/CopyCasusCodePopup.js:
```diff
@@ -0,0 +1,95 @@
+							{copyButton}{cancelButton}
+						</div>
+					</Popup>
+					<ToastContainer />
```
We probably want this container just once on the outermost layer, right?

SecondThread commented on src/tanks/Tank.js:
```diff
@@ -106,7 +106,7 @@ class Tank extends GameObject {
 	// Casus:
 	interpriterState: InterpriterState;
-	casusCode: CasusBlock;
+	casusCode: ContainerBlock;
```
Yeah, all of these Tank changes look substantially fine to me. As long as they pass flow that means we have been doing things consistently and it shouldn't be a problem for us to make decisions based on this being a ContainerBlock.

SecondThread reviewed (changes requested):
Mostly good. The two substantial changes are:
- we already have an api call for setting the casus code and we probably don't want a second one. (plus the existing one is really easy to modify to do what you need this to, so you should use that instead)
- setTankBeingCopiedFrom is unnecessarily long and complicated

Zexdur committed:
Worked on suggestions

Zexdur committed:
Fixed flow

Zexdur committed:
Merge branch 'master' into CasusFromOneTankToAnother

Zexdur committed:
fixed warnings

SecondThread reviewed (changes requested):
The code looks fine. From a user perspective it needs work:
- The button is unreasonably large. It spans 3 lines compared to other 3 buttons next to it, which are all one line:
![image](https://user-images.githubusercontent.com/18317476/79503386-12fa4100-7fff-11ea-9e0e-89e1c8b19683.png)
- The button is in the tank-options section, but it is a casus option, so it would make more sense to be by the casus button.
- The window doesn't close when you click copy.
- There is no confirmation asking the user if they are sure (this will overwrite all casus code in the previous tank). You could just delete a very large amount of work here if you select the wrong tank. You could implement this by needing to check a box saying "I understand this will overwrite all code in [whatever tank]" before the copy button is enabled

Zexdur committed:
Added some user things in the copy code branch

Zexdur committed:
Update CopyCasusCodePopup.js

Closed popup once casus is saved

SecondThread reviewed (changes requested):
- The tank that you have selected in shown in the list which is confusing
- Just disable the "copy" button when the checkbox isn't clicked, don't show the user that they can click it and then give them an error when they do
- The Copy Code button should be light blue. Otherwise we have two primary buttons on one page.
- It is more clear if you use the word "overwriting" instead of "rewriting". You aren't writing the code again, you are deleting it and replacing it with something else
- Even if you were going to use the word "rewriting", you should spell it "rewriting" instead of "rewritting"

Zexdur committed:
Update CopyCasusCodePopup.js

Only have to handle removing current tank from drop down but getting weird bug when I try to do splice. Will try again later.

Zexdur committed:
Update CopyCasusCodePopup.js

Made so that the current tank is not in the list of dropdowns

SecondThread commented on src/armory/Armory.js:
```diff
@@ -324,10 +325,18 @@ class Armory extends React.Component<Props, State> {
-					<label>&emsp;Edit Code</label>
+					<br/><br/>
+					{selectedTank==null?<div></div>:
+							<CopyCasusCodePopup
```
fix indentation here, one too many tabs

SecondThread commented on src/armory/CopyCasusCodePopup.js:
```diff
@@ -0,0 +1,120 @@
+		const cancelButton = (
+			<button className="cancelbtn" onClick={() => this.setState({popupOpen: false})}>Cancel</button>
+		);
+		const copyButton = this.state.confirmed===false?(<button className="smallbtn" disabled>Copy</button>):(<button className="smallbtn" onClick={() => this.copyCasusCode()}>Copy</button>);
```
also make sure that they have selected a tank before enabling

SecondThread commented on src/casus/saveCasus.js:
```diff
@@ -5,9 +5,12 @@ import getTankForCasus from '../globalComponents/getTankForCasus.js';
+function saveCasus(casusCode: ContainerBlock, tankID: ?string, onLoad:() => void): void {
 	
-	const tankToEditID: string = getTankForCasus();
+	let tankToEditID: ?string = tankID;
```
You can clean this up to be: `const tankToEditID = tankID ?? getTankForCasus();`

SecondThread reviewed (approved):
Please fix changes before merging

Zexdur committed:
Fixed Suggestions
